### PR TITLE
 feat(ws1): NativeSiLUOp + NativeSwiGLUOp pure-PyTorch ground-truth references + numerical contract tests

### PR DESCRIPTION
**Pull request by maxiaosong1124**
## Summary

Adds the pure-PyTorch ground-truth reference ops for the **gated MLP activation**
of the WS1 batch-invariant forward chain: **SiLU** (Swish) and **SwiGLU**, built on
top of the numerical contract defined in #108. Ships the two ops, their registry
wiring, docs, and a 16-case test suite that pins down both alignment axes
(Axis-A bitwise batch invariance, Axis-B per-dtype path).

Refs #108

## Terminology

This PR uses the WS1 alignment vocabulary from #108:

- **Axis-A — batch invariance (reproducibility).** A row's output must not depend on
  how many rows share the batch (batch size, slicing, padding). Asserted **bitwise**
  (`torch.equal`). This is what keeps train-time (large batch) and sample-time
  (small batch / dynamic padding) numerics identical so the policy ratio doesn't drift.
- **Axis-B — accuracy.** The low-precision (bf16 / fp16) forward path. These activations
  compute in fp32 and cast the result back to the input dtype, so the dtype output is
  **bitwise** equal to the fp32 formula cast down — asserted with `torch.equal` against
  an independent fp32 reference (no tolerance window needed for an element-wise op).

## Motivation / Context

#108 establishes the ground-truth harness and numerical contract for the WS1
batch-invariant forward chain. The Qwen3-8B dense MLP is a gated (SwiGLU) MLP:

down_proj( silu(gate_proj(x)) * up_proj(x) )

This PR covers the activation stage in the middle:

- **`silu`** — element-wise `x * sigmoid(x)` (`hidden_act="silu"`), shape-agnostic.
- **`swiglu`** — `silu(gate) * up`, where `gate` / `up` are the gate_proj / up_proj
  outputs at the intermediate dim (Qwen3-8B: **12288**). The trailing `down_proj` is a
  plain matmul and lives in a separate op.

This PR provides the deterministic fp32 reference path those downstream kernels
(Triton / CUDA / ROCm fused activation) will be validated against.

## Changes

- **`rl_engine/kernels/ops/pytorch/activation/swiglu.py`** — `NativeSiLUOp`, `NativeSwiGLUOp`
  - `forward()` — accumulate in fp32, cast result back to input dtype (Axis-B path)
  - `forward_fp32()` — fp32 accumulation, forced fp32 output (ground-truth / backward golden source)
  - Formulas: `silu(x) = x * sigmoid(x)`; `swiglu(gate, up) = gate * sigmoid(gate) * up`
  - Pure functions — inputs never mutated in place
  - SwiGLU shape guard: `gate` and `up` must share shape
- **`rl_engine/kernels/registry.py`** — register `PYTORCH_NATIVE_SILU` /
  `PYTORCH_NATIVE_SWIGLU` in `OpBackend` and add `silu` / `swiglu` dispatch to the
  cuda / rocm / cpu priority maps
- **`tests/test_swiglu.py`** — 16 tests (details below)
- **`docs/operators/activation.md`** + nav / index wiring

## How this satisfies the #108 contract

| #108 requirement | How it's met here |
|---|---|
| Deterministic reference path | `forward_fp32()` computes element-wise in fp32; tests use fixed-seed `torch.Generator` so outputs are reproducible |
| Per-dtype tolerance policy (bitwise vs tight-tolerance) | **Axis-A** asserted **bitwise** (`torch.equal`); **Axis-B** dtype output is fp32-compute-then-cast, asserted **bitwise** against the independent fp32 formula cast to dtype (element-wise op needs no tolerance window) |
| Batch-config sweep / validation helper | Batch-invariance checks compute on the full batch, then assert sliced/padded rows are **bitwise identical** to their full-batch counterparts |
| Realistic shapes covered | Batch-invariance tests run at the Qwen3-8B intermediate dim (`12288`) |

## Test Environment

| | |
|---|---|
| OS | Ubuntu 22.04.5 LTS (kernel 5.15.0-122-generic) |
| Python | 3.12.3 |
| PyTorch | 2.8.0+cu128 |
| CUDA / cuDNN | 12.8 / 9.10.02 (driver 580.65.06) |

──────────────────────────
padding variants, asserted bitwise
- purity (inputs not mutated in place)
- gradient flow (fp32 autograd = backward golden source)
- SwiGLU shape guard fires on mismatched gate / up shapes
- registry dispatch resolves silu → NativeSiLUOp, swiglu → NativeSwiGLUOp

Checklist

- [x] Pure-PyTorch reference, no custom extension required
- [x] SwiGLU covered at the Qwen3-8B intermediate dim (12288)
- [x] Axis-A bitwise batch invariance enforced
- [x] Axis-B fp32-compute-then-cast dtype path tested
- [x] Registered in OpBackend + cuda/rocm/cpu priority maps
- [x] All 16 tests pass locally

---

## Summary by CodeRabbit

* **New Features**
  * Added SiLU and SwiGLU activation operators with PyTorch implementations and registry support.

* **Documentation**
  * Added comprehensive documentation for SiLU/SwiGLU operators, including mathematical definitions, tensor shape requirements, and backend dispatch behavior.
  * Added validation tests covering correctness, input validation, batch invariance, and gradient propagation.

### CODE REVIEW

File: tests/test_swiglu.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    assert torch.isfinite(x.grad).all()
+
+
+def test_swiglu_gradient_flows():
+    op = NativeSwiGLUOp()
+    gate = _rand((2, _INTERMEDIATE), seed=12).requires_grad_(True)
+    up = _rand((2, _INTERMEDIATE), seed=13).requires_grad_(True)
+    op.forward_fp32(gate, up).sum().backward()
+    assert torch.isfinite(gate.grad).all() and torch.isfinite(up.grad).all()
+
```
Comment (Flink-ddd) [inline tests/test_swiglu.py:123]: Currently, test_silu_gradient_flows and test_swiglu_gradient_flows only check if the gradients are numerically valid (isfinite). To strictly satisfy the WS1 contract, we must use a slice test to prove that the backward pass (gradient computation) for both SiLU and SwiGLU is completely independent of the batch size.

Code Example (for SwiGLU):

```
# Add this to tests/test_swiglu.py
def test_swiglu_backward_batch_invariance_slice():
    """Axis A: Gradients must be bitwise identical regardless of batch size."""
    op = NativeSwiGLUOp()
    
    # 1. Full batch forward & backward
    gate_full = _rand((8, 32, _INTERMEDIATE), seed=1).requires_grad_(True)
    up_full = _rand((8, 32, _INTERMEDIATE), seed=2).requires_grad_(True)
    out_full = op.forward_fp32(gate_full, up_full)
    
    dy_full = _rand(out_full.shape, seed=3)
    out_full.backward(dy_full)
    
    grad_gate_full_sliced = gate_full.grad[:1].clone()
    grad_up_full_sliced = up_full.grad[:1].clone()
    
    # 2. Sliced batch (batch size = 1)
    gate_slice = _rand((8, 32, _INTERMEDIATE), seed=1)[:1].detach().requires_grad_(True)
    up_slice = _rand((8, 32, _INTERMEDIATE), seed=2)[:1].detach().requires_grad_(True)
    out_slice = op.forward_fp32(gate_slice, up_slice)
    
    # Use matching slice of upstream grad
    out_slice.backward(dy_full[:1])
    
    # 3. Assert gradients are bitwise identical
    assert torch.equal(gate_slice.grad, grad_gate_full_sliced)
    assert torch.equal(up_slice.grad, grad_up_full_sliced)

# Note: Please add a similar test for SiLU (test_silu_backward_batch_invariance_slice)
```

File: rl_engine/kernels/ops/pytorch/activation/swiglu.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    """
+    Pure PyTorch native SiLU reference.
+    out = x * sigmoid(x)   (a.k.a. Swish)
+
+    Element-wise activation used by the Qwen3 SwiGLU MLP
+    (hidden_act="silu"). No hyper-parameters and shape-agnostic, so the
+    Qwen3-8B intermediate dim (12288) is just one valid last-dim size.
+    """
+
+    def __init__(self) -> None:
```
Comment (Flink-ddd) [inline rl_engine/kernels/ops/pytorch/activation/swiglu.py:20]: Just like the RMSNorm op, these pure PyTorch reference operators should inherit from nn.Module to ensure seamless integration with the broader PyTorch ecosystem (Hooks, state_dict, Dynamo tracing) later down the line.
Bonus: Because nn.Module automatically handles routing __call__ to forward, you can safely delete the manually defined __call__ methods in both classes, making the code even cleaner.

**Review by Flink-ddd — CHANGES_REQUESTED** [review]
This PR maintains the same high engineering standards as #160. Writing out the explicit mathematical formulas (like x * sigmoid(x)) is excellent documentation for whoever builds the downstream CUDA/Triton fused kernels. Since these two PRs were likely developed in parallel, this one shares the exact same two minor omissions.

**Review by KJLdefeated — APPROVED** [review]
Math and structure are correct, dual-path contract is consistent with the rest of WS1, and Vensen's two points cover most of what I'd raise. Pls resolve Vensen's requests. LGTM.

**Comment by maxiaosong1124** [comment]
@Flink-ddd @KJLdefeated Thanks for the careful review! All requested changes have been pushed in 2b3db46. Summary of what was addressed:
both points resolved:

1. nn.Module inheritance — NativeSiLUOp and NativeSwiGLUOp now inherit from nn.Module (with super().__init__()), matching the RMSNorm op. As you noted, the manual __call__ methods are now removed since nn.Module routes __call__ → forward automatically.
2. Backward batch-invariance slice tests — Added test_silu_backward_batch_invariance_slice and test_swiglu_backward_batch_invariance_slice. Instead of only checking isfinite, they now assert (via torch.equal) that the gradients of the batch-size-1 slice are bitwise identical to the corresponding slice of the full-batch gradients — i.e. the Axis-A backward contract, following the example you provided.

**Review by Flink-ddd — APPROVED** [review]
LGTM now, Thank you for update.

**Comment by Flink-ddd** [comment]
Let's merge this PR first
